### PR TITLE
Accumulated changes from gerasdf's fork

### PR DESCRIPTION
**Pull request by gerasdf**
A quick Changelog

* `bot.getToken()`
* refactored common code (less error prone, smaller FLASH footprint)
* simplified Json6 usage (smaller RAM footprint)
* reduced String usage of RAM
* added reply_to support (bot now can follow a conversation)
* fixed bugs on testing empty Strings
* id of last message sent available to user (`bot.last_sent_message_id`)
* user can (dynamically) configure `bot.maxMessageLength`
* implemented `bot.answerCallbackQuery()` to reply to inline keyboard callbacks

and it also includes merged contributions from other forks
* `bot.updateToken()` by rafaelcbarbosa
* `bot.setMyCommands()` by didgerihorn
* `bot.getFile()` attachments and OTA by solcer

examples for `bot.answerCallbackQuery()`, `bot.maxMessageLength`, ``bot.last_sent_message_id`, `bot.getToken()`, `bot.updateToken()` and `reply_to_*` are missing. I will work on them on a separate fork starting from this commit, and then do another pull request for the examples. Let me know if you think this is a good idea

### CODE REVIEW

File: src/UniversalTelegramBot.cpp [hunk 1]
```diff
@@ -658,90 +669,88 @@ String UniversalTelegramBot::sendPostPhoto(JsonObject &payload) {
...
+                                       const String& keyboard) {
 
+  DynamicJsonDocument payload(maxMessageLength);
   payload["chat_id"] = chat_id;
   payload["photo"] = photo;
 
-  if (caption) {
-    payload["caption"] = caption;
-  }
+  if (!caption.isEmpty())
```
Comment (witnessmenow) [inline src/UniversalTelegramBot.cpp:701]: This is working for me? I also don't see it on the Arduino String docs

https://www.arduino.cc/reference/en/language/variables/data-types/stringobject/

File: src/UniversalTelegramBot.cpp [hunk 2]
```diff
@@ -658,90 +669,88 @@ String UniversalTelegramBot::sendPostPhoto(JsonObject &payload) {
...
+      payload["disable_notification"] = disable_notification;
 
-  if (keyboard) {
-    JsonObject &replyMarkup = payload.createNestedObject("reply_markup");
+  if (reply_to_message_id && reply_to_message_id != 0)
+      payload["reply_to_message_id"] = reply_to_message_id;
 
-    DynamicJsonBuffer keyboardBuffer;
-    replyMarkup["keyboard"] = keyboardBuffer.parseArray(keyboard);
+  if (!keyboard.isEmpty()) {
```
Comment (witnessmenow) [inline src/UniversalTelegramBot.cpp:710]: Same

File: src/UniversalTelegramBot.cpp [hunk 3]
```diff
@@ -751,9 +760,49 @@ bool UniversalTelegramBot::sendChatAction(String chat_id, String text) {
...
+}
+
+bool UniversalTelegramBot::answerCallbackQuery(const String &query_id, const String &text, bool show_alert, const String &url, int cache_time) {
+  DynamicJsonDocument payload(maxMessageLength);
+
+  payload["callback_query_id"] = query_id;
+  payload["show_alert"] = show_alert;
+  payload["cache_time"] = cache_time;
+
+  if (!text.isEmpty()) payload["text"] = text;
```
Comment (witnessmenow) [inline src/UniversalTelegramBot.cpp:797]: same

File: src/UniversalTelegramBot.cpp [hunk 4]
```diff
@@ -751,9 +760,49 @@ bool UniversalTelegramBot::sendChatAction(String chat_id, String text) {
...
+  if (!url.isEmpty()) payload["url"] = url;
```
Comment (witnessmenow) [inline src/UniversalTelegramBot.cpp:798]: same

**Comment by witnessmenow** [comment]
Echobot works after changing the 4 things above to

`
//if(text.isEmpty())
if(text)
`

I assume it still works as intended!

**Comment by gerasdf** [comment]
The test kind of fails, I assume you want `""` (empty String) to be false, but:

```C++
void test() {
  String a;
  String b = "";
  String c = "hola";
  
  if (a) Serial.println("a is true");
  if (b) Serial.println("b is true");
  if (c) Serial.println("c is true");
  if (!a.isEmpty()) Serial.println("a.isEmpty() is false");
}
```

prints

```
a is true
b is true
c is true
```

which kind of defeats the purpose (and `!isEmpty()` works apparently fine)

**Comment by witnessmenow** [comment]
Ok fair enough that if(string) doesn't work as i expected, but isEmpty
doesn't compile for me. What version of Arduino are you using?

**Comment by gerasdf** [comment]
I'm on arduino-1.8.12, no special setup AFAIK

**Comment by witnessmenow** [comment]
On 1.8.5 on a mac and it doesn't compile

img

**Comment by witnessmenow** [comment]
I suppose `if(string.length() > 0)` should work?

**Comment by gerasdf** [comment]
Sure! Or `> 0`

**Comment by gerasdf** [comment]
I don't remember why, but I was sort of forced to upgrade from 1.8.5. (it wasn't just for this though)

**Pull request merged**